### ISSUE TITLE
Vnode startup can hang the vnode_manager in pmap

### ISSUE THREAD

**Issue by bearcage** [basho/riak_core#662]
Reported by @angrycub & @joecaswell

When handling get_vnode calls, vnode_manager calls pmap over all vnodes with a start fun which calls riak_core_vnode:wait_for_init, which is an infinity-timeout gen_fsm sync_send_event. If a vnode hangs in init, the pmap collect will never be acked, and the pending set will never empty, which causes the pmap to loop in collect.

Observed in action at a customer site, resolved by deleting AAE trees to allow the stuck (AAE) vnode to start.

**Comment by engelsanchez** [comment]
Given that a vnode may take hours (heck, we have seen days in pathological cases) to start, it wouldn't be wise to change this to any kind of timeout. It is reasonable for the mentioned code to wait forever as long as the vnode is alive and trying to start up. The code on this layer relies on the vnode not being bug ridden and stuck forever on something.  It sounds like a proper fix would have to happen in the vnode startup code, closer to the root cause.

**Comment by bearcage** [comment]
Agreed; I don't like the notion of a timeout at all since even a proper startup can take quite a long time for Bitcask clusters with lots of data to load and no hintfiles available. Seems to me we might be able to mitigate this in the AAE case a little by adding a timeout on the aae-vnode side, and just blowing away the vnode's data if that fails.

**Comment by seancribbs** [comment]
Even if a vnode can take forever to startup, it doesn't negate the fact that this problem blocks the vnode manager from processing other messages when a single pmap response is not received. I would have suggested using a link or monitor, but `spawn_link` is already used to initiate the workers. Because this happens in-process, you can't isolate those exits from other `'EXIT'` or `'DOWN'` messages that the vnode manager might receive. The solution is really another process.

**Comment by seancribbs** [comment]
The alternative would be to inject a `try/catch` into the pmap worker, and send error terms to the caller in the result, but that would likely allow things to fail silently, which we probably don't want.

**Comment by engelsanchez** [comment]
Yes. if you are talking about the vnode manager, it should be reworked to not allow this blocking process from blocking the manager itself. If you think about it, a lot of code indirectly blocks on this everywhere. Once you start pulling the threads, it will go a long way in many directions.  Where is my Core 3.0?
